### DISCUSSION TITLE
The command "node ^"D:^\Xampp^\htdocs^\Project^\vendor^\spatie^\browsershot^\src/../bin/browser.js^" ^"^{^\^"url^\^":^\^"https:^\/^\/example.com^\^",^\^"action^\^":^\^"screenshot^\^",^\^"options^\^":^{^\^"type^\^":^\^"png^\^",^\^"args^\^":^[^],^\^"viewport^\^":^{^\^"width^\^":800,^\^"height^\^":600^}^}^}^"" failed. Exit Code: 1(General error) Working directory: D:\Xampp\htdocs\Project\public Output: ================ Error Output: ================ 'node' is not recognized as an internal or external command, operable program or batch file.

### DISCUSSION THREAD

**Discussion by jerry4rahul** [spatie/browsershot#606] [Q&A]
`The command "node ^"D:^\Xampp^\htdocs^\Project^\vendor^\spatie^\browsershot^\src/../bin/browser.js^" ^"^{^\^"url^\^":^\^"https:^\/^\/example.com^\^",^\^"action^\^":^\^"screenshot^\^",^\^"options^\^":^{^\^"type^\^":^\^"png^\^",^\^"args^\^":^[^],^\^"viewport^\^":^{^\^"width^\^":800,^\^"height^\^":600^}^}^}^"" failed. Exit Code: 1(General error) Working directory: D:\Xampp\htdocs\Project\public Output: ================ Error Output: ================ 'node' is not recognized as an internal or external command, operable program or batch file. `

The error looks like node is not installed but by use `node -v` returns v17.6.0

Can anyone help me on this?

**Comment by andv95** [comment]
I also had the same problem, @jerry4rahul  did you solve it?

**Reply by jerry4rahul** [reply]
No @andv95

**Reply by andv95** [reply]
I resloved it. May be APP_URL was 127.0.0.1:8000, i think you should be set vitual host for project.

**Reply by Adamcy-lvts** [reply]
I am using a virtual host using wamp server still this error persists

**Reply by abbasmashaddy72** [reply]
You Need to Set Node & NPM Binary location & for sure error will be resolved

also, you need to install npm package puppeteer required for the package to work properly

**Comment by abbasmashaddy72** [comment]
Same Issue but solved by adding
`->setNodeBinary('/usr/bin/node')
            ->setNpmBinary('/usr/bin/npm')`

and installing `npm i puppeteer`

**Reply by SaddamDeveloper** [reply]
I have the same issue. can anyone tell me how it is fixed in the windows machine?

**Reply by prageeth-peiris** [reply]
I fixed this by running laravel app on apache virtual host.
Refer : https://stackoverflow.com/questions/27268205/how-to-create-virtual-host-on-xampp

**Comment by tuarrep** [comment]
The command escaping is broken, see https://github.com/spatie/browsershot/blob/main/src/Browsershot.php#L897

The whole command is escaped including the node binary path which shouldn't be.

I'll try to take some spare time to fix and test this, if no-one did it before me.

**Reply by JBtje** [reply]
Lucky you... someone did it before...
https://web.archive.org/web/20201031143934/https://github.com/spatie/browsershot/issues/271#issuecomment-480772391
_Why where issues removed from this project!? It took me an hour to re-find this using web archive..._

Current working implementation on win 10:
**browsershot.php**
```
    protected function createTemporaryOptionsFile( string $command ): string
    {
        $this->temporaryOptionsDirectory = (new TemporaryDirectory( $this->tempPath ))->create();

        file_put_contents( $temporaryOptionsFile = $this->temporaryOptionsDirectory->path( 'command.js' ), $command );

        return "file://C:{$temporaryOptionsFile}"; // not sure if this is needed, but node might require a full path
        //return "file://:{$temporaryOptionsFile}";
    }
...

    protected function getFullCommand( array $command )
    {
...
        if( $this->isWindows() ) {
            if( $this->writeOptionsToFile ) {
                $temporaryOptionsFile  = $this->createTemporaryOptionsFile( json_encode( $command ) );
                $wronglyEscapedCommand = '-f ' . $temporaryOptionsFile;
            }
            else {
                $wronglyEscapedCommand = json_encode( $command );
            }

            return escapeshellarg( $nodeBinary ) . ' '
                . escapeshellarg( $binPath ) . ' '
                . '"' . base64_encode( $wronglyEscapedCommand ) . '"';
        }
```
Note: The `escapeshellcmd` caused me trouble when using the full path to `node`, so i just removed it and escape the `$nodebin`, security wise this _might_ not be best... (i only use this local)

**browser.js**
```
var atob = require('atob');

const fs = require('fs');
const URL = require('url').URL;
const URLParse = require('url').parse;

const [, , ...args] = process.argv;
args[0] = atob( args[0] ) // ADD THIS LINE, below the above line that already is in the file, this will break Unix however...
...
```
By the time I figured out how to make it work on windows, I didn't have energy left to get the PR going...
You're free to make this into a PR. but please check security implementations of the code, I only use this locally and didn't bother to check consequences of the code.
Possibly you can even get ride of the base64 encode, so that `atob` is not needed in browser.js. Otherwise you somehow have to check in browser.js if args[0] is base64 encoded (check is now missing) or it won't work on Unix (i.g. by adding another parameter to browser.js).
